packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts

```diff
@@ -0,0 +1,427 @@
+		]);
+		await writeAsync(dbSocket, sslRequestPacket);
+
+		// Upgrade server connection to TLS
+		const tlsOptions: tls.ConnectionOptions = {
+			socket: dbSocket,
+			host: targetHost,
+			servername: targetHost,
+			minVersion: "TLSv1.2",
+			rejectUnauthorized: true,
+		};
+
+		try {
+			const tlsSocket = await tlsConnect(tlsOptions);
+
+			// Send original init packet to client (with SSL capability)
+			const fullInitPacket = Buffer.concat([
+				initPacketChunk.subarray(0, 4),
+				fullPayloadBody,
+			]);
+			await writeAsync(clientSocket, fullInitPacket);
+			const clientAuthPayload = await readAsync(clientSocket);
+
+			// Forward client auth to server and increment sequence ID byte to keep client driver/server in sync
+			clientAuthPayload[3]++;
+			await writeAsync(tlsSocket, Buffer.from(clientAuthPayload));
+
+			// Read and forward server's response, and fix sequence ID byte back to client
+			const authResponsePayload = await readAsync(tlsSocket);
+			authResponsePayload[3]--;
+			await writeAsync(clientSocket, Buffer.from(authResponsePayload));
+
+			// Set up pipes with error handler
+			setupTLSConnection(clientSocket, tlsSocket);
+			return;
+		} catch (e) {
+			if (!sslmodeRequire) {
+				throw e;
+			}
+			// Attempt to fall back to plain TCP
+		}
+	}
+	if (sslmodeRequire) {
+		throw new Error("Server SSL failed, but client requires SSL.");
+	}
+	// Disconnect original socket to DB
+	dbSocket.destroy();
+
+	// Create new connection and send init packet without SSL capability
+	const newDbSocket = await createPlainTCPConnection(
+		targetHost,
+		targetPort,
+		clientSocket
+	);
+
+	// Pipe plain TCP sockets
+	clientSocket.pipe(newDbSocket);
+	newDbSocket.pipe(clientSocket);
+	return;
+}
+
+/** Helper function to create and wait for plain TCP connection */
+async function createPlainTCPConnection(
+	targetHost: string,
+	targetPort: number,
+	clientSocket: net.Socket
+): Promise<net.Socket> {
+	const dbSocket = net.connect({ host: targetHost, port: targetPort });
+
+	// Wait for connection to be established
+	await new Promise<void>((resolve, reject) => {
+		const handleConnect = () => {
+			dbSocket.off("error", handleError);
+			resolve();
+		};
+
+		const handleError = (err: Error) => {
+			dbSocket.off("data", handleConnect);
+			reject(err);
+		};
+
+		dbSocket.once("connect", handleConnect);
+		dbSocket.once("error", handleError);
+	});
+
+	// Set up error handler
+	dbSocket.on("error", () => {
+		clientSocket.destroy();
+	});
+
+	return dbSocket;
+}
+
+/** Helper function to set up TLS connection with pipes and error handlers */
+function setupTLSConnection(
+	clientSocket: net.Socket,
+	tlsSocket: tls.TLSSocket
+): void {
+	// Set up error handler for runtime TLS errors
+	tlsSocket.on("error", () => {
+		clientSocket.destroy();
+		tlsSocket.destroy();
+	});
+
+	// Pipe sockets
+	clientSocket.pipe(tlsSocket);
+	tlsSocket.pipe(clientSocket);
+}
+
+/** Helper functions for net library */
+function writeAsync(
```

Comment:
suggestion, it looks like this function is often called with `Buffer.from(array)`, would it makes sense that bytes could also be an array and turned into a Buffer here if needed?